### PR TITLE
crimson/osd, common: implement the inject{m,}dataerr admin commands

### PR DESCRIPTION
**Pull request by rzarzynski**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/common/cmdparse.h [hunk 1]
```diff
@@ -94,6 +107,19 @@ bool cmd_getval(
   }
 }
 
+template <typename T>
+std::optional<T> cmd_getval(const cmdmap_t& cmdmap,
+			    const std::string& k,
```
Comment (tchaikov) [inline src/common/cmdparse.h:112]: might want to use `std::string_view` ?

File: src/crimson/admin/osd_admin.cc [hunk 2]
```diff
@@ -265,4 +272,96 @@ class SeastarMetricsHook : public AdminSocketHook {
...
+{
+  int64_t pool;
+  if (auto poolarg = cmd_getval<std::string>(cmdmap, "pool"); poolarg) {
+    pool = osdmap.lookup_pg_pool_name(*poolarg);
+    if (pool < 0 && std::isdigit((*poolarg)[0])) {
+      pool = std::atoll(poolarg->c_str());
+    }
+    if (pool < 0) {
+      // the return type of `fmt::format` is `std::string`
+      throw fmt::format("Invalid pool '{}'", *poolarg);
```
Comment (tchaikov) [inline src/crimson/admin/osd_admin.cc:288]: nit, how about using `std::invalid_argument` or `crimson::osd::invalid_argument`? i am inclined to use classes inherited from `exception` for exceptions.

Comment (tchaikov) [reply src/crimson/admin/osd_admin.cc:288]: also, could use
```c++
"Invalid pool '{}'"_format(*poolarg);
```
for compile time fmt parameter validation.

File: src/crimson/admin/osd_admin.cc [hunk 3]
```diff
@@ -265,4 +272,96 @@ class SeastarMetricsHook : public AdminSocketHook {
...
+    }
+  } else {
+    throw std::string{"No 'pool' specified"};
+  }
+
+  std::string objname, nspace;
+  pg_t rawpg;
+  if (auto objarg = cmd_getval<std::string>(cmdmap, "objname"); objarg) {
+    if (std::size_t sep_pos = objname.find_first_of('/');
+        sep_pos != std::string::npos) {
```
Comment (tchaikov) [inline src/crimson/admin/osd_admin.cc:298]: s/std::string::npos/objname.npos/ for better readability.

File: src/crimson/admin/osd_admin.cc [hunk 4]
```diff
@@ -265,4 +272,96 @@ class SeastarMetricsHook : public AdminSocketHook {
 };
 template std::unique_ptr<AdminSocketHook> make_asok_hook<SeastarMetricsHook>();
 
+
+static ghobject_t test_ops_get_object_name(
+  const OSDMap& osdmap,
+  const cmdmap_t& cmdmap)
+{
+  int64_t pool;
+  if (auto poolarg = cmd_getval<std::string>(cmdmap, "pool"); poolarg) {
```
Comment (tchaikov) [inline src/crimson/admin/osd_admin.cc:281]: might want to early return for error handling, like
```c++
auto pool = [&] {
  cmd_getval<std::string>(cmdmap, "pool");
  if (!pool_arg) {
    // throw
  }
  int64_t pool = osdmap.lookup_pg_pool_name(*poolarg);
  // more error handling
  return pool;
}();
```

File: src/common/cmdparse.h (hunk 1 above)
Comment (rzarzynski) [reply src/common/cmdparse.h:112]: Well, this is just a wrapper which requires `std::string`.

File: src/crimson/admin/osd_admin.cc (hunk 4 above)
Comment (rzarzynski) [reply src/crimson/admin/osd_admin.cc:281]: Not sure whether extra lambda improves readability. Anyway, I have no strong feelings here.

Comment (tchaikov) [reply src/crimson/admin/osd_admin.cc:281]: i think the lambda is not the key to improve the readability here. the point is

- to reduce the indent level
- to highlight the happy path

File: src/common/cmdparse.h (hunk 1 above)
Comment (tchaikov) [reply src/common/cmdparse.h:112]: okay, will create a PR to fix this.

File: src/crimson/admin/osd_admin.cc [hunk 5]
```diff
@@ -265,4 +272,141 @@ class SeastarMetricsHook : public AdminSocketHook {
...
+    if (object_locator_t oloc(pool, nspace);
+        osdmap.object_locator_to_pg(object_t(objname), oloc,  rawpg) < 0) {
+      throw std::invalid_argument{"Invalid namespace/objname"};
+    }
+  } else {
+    throw std::invalid_argument{"No 'objname' specified"};
+  }
+
+  int64_t shardid;
+  if (auto shardidarg = cmd_getval<int64_t>(cmdmap,
```
Comment (tchaikov) [inline src/crimson/admin/osd_admin.cc:314]: by introducing the lambda, could remind us to set `shardid` here.

File: src/crimson/admin/osd_admin.cc [hunk 6]
```diff
@@ -265,4 +272,141 @@ class SeastarMetricsHook : public AdminSocketHook {
...
+      throw std::invalid_argument{
+	fmt::format("Invalid pool '{}'"_format(*poolarg))
+      };
+    }
+  } else {
+    throw std::invalid_argument{"No 'pool' specified"};
+  }
+
+  std::string objname, nspace;
+  pg_t rawpg;
```
Comment (tchaikov) [inline src/crimson/admin/osd_admin.cc:298]: s/rawpg/raw_pg/

File: src/crimson/admin/osd_admin.cc [hunk 7]
```diff
@@ -265,4 +272,141 @@ class SeastarMetricsHook : public AdminSocketHook {
...
+    }
+    if (object_locator_t oloc(pool, nspace);
+        osdmap.object_locator_to_pg(object_t(objname), oloc,  rawpg) < 0) {
+      throw std::invalid_argument{"Invalid namespace/objname"};
+    }
+  } else {
+    throw std::invalid_argument{"No 'objname' specified"};
+  }
+
+  int64_t shardid;
```
Comment (tchaikov) [inline src/crimson/admin/osd_admin.cc:313]: s/shardid/shard_id/

File: src/crimson/admin/osd_admin.cc [hunk 8]
```diff
@@ -265,4 +272,141 @@ class SeastarMetricsHook : public AdminSocketHook {
...
+	fmt::format("Invalid pool '{}'"_format(*poolarg))
+      };
+    }
+  } else {
+    throw std::invalid_argument{"No 'pool' specified"};
+  }
+
+  std::string objname, nspace;
+  pg_t rawpg;
+  if (auto objarg = cmd_getval<std::string>(cmdmap, "objname"); objarg) {
```
Comment (tchaikov) [inline src/crimson/admin/osd_admin.cc:299]: s/objarg/obj_arg/

File: src/crimson/admin/osd_admin.cc [hunk 9]
```diff
@@ -265,4 +272,141 @@ class SeastarMetricsHook : public AdminSocketHook {
...
+    if (std::size_t sep_pos = objname.find_first_of('/');
```
Comment (tchaikov) [inline src/crimson/admin/osd_admin.cc:300]: might want to set `objname`?

File: src/common/cmdparse.h (hunk 1 above)
Comment (tchaikov) [reply src/common/cmdparse.h:112]: #41434

File: src/crimson/admin/osd_admin.cc [hunk 10]
```diff
@@ -265,4 +272,111 @@ class SeastarMetricsHook : public AdminSocketHook {
...
+    }
+    int64_t pool = osdmap.lookup_pg_pool_name(*pool_arg);
+    if (pool < 0 && std::isdigit((*pool_arg)[0])) {
+      pool = std::atoll(pool_arg->c_str());
+    }
+    if (pool < 0) {
+      // the return type of `fmt::format` is `std::string`
+      using namespace fmt::literals;
+      throw std::invalid_argument{
+        fmt::format("Invalid pool '{}'"_format(*pool_arg))
```
Comment (tchaikov) [inline src/crimson/admin/osd_admin.cc:293]: no need to use both `fmt::format()` and `operator "" _format()` .

**Review by tchaikov — APPROVED** [review]

**Comment by tchaikov** [comment]
@rzarzynski since #41434 was merged, i rebased your change on top of master to resolve the conflicts so it can be merged sooner.

**Pull request merged**